### PR TITLE
Statics

### PR DESCRIPTION
**Pull request by Michele-Martini**
Tentative approach to solve the statics of the system by solving the root finding problem of the statics equation with Newton-Raphson solver.

### CODE REVIEW

File: src/soromox/systems/pendulum.py [hunk 1]
```diff
@@ -692,6 +692,50 @@ def forward_dynamics(
         qdd = jnp.linalg.solve(B, rhs)
         return jnp.concatenate([qd, qdd])
 
+    @eqx.filter_jit
+    def statics_residual(
```
Comment (mstoelzle) [inline src/soromox/systems/pendulum.py:696]: Maybe for consistency, this could be renamed to `forward_statics`? Is this an expression that is sometimes used?

File: src/soromox/systems/pendulum.py [hunk 2]
```diff
@@ -692,6 +692,50 @@ def forward_dynamics(
...
+
+        if u is None:
+            u = jnp.zeros((self.num_actuators,))
+        if tau_ext is None:
+            tau_ext = jnp.zeros((self.num_links,))
+
+        G = self.gravitational_force(q)
+        tau_el = self.elastic_force(q)
+        tau_u = self.actuation_force(q, u)
+        residual = tau_el + G - tau_u - tau_ext
```
Comment (mstoelzle) [inline src/soromox/systems/pendulum.py:735]: I think it would make sense to switch the sign here to match the convention of the forward dynamics.

File: src/soromox/systems/pendulum.py [hunk 3]
```diff
@@ -692,6 +692,50 @@ def forward_dynamics(
...
+            raise ValueError("actuation_args must be None, (u,), or (u, tau_ext)")
+
+        if u is None:
+            u = jnp.zeros((self.num_actuators,))
+        if tau_ext is None:
+            tau_ext = jnp.zeros((self.num_links,))
+
+        G = self.gravitational_force(q)
+        tau_el = self.elastic_force(q)
+        tau_u = self.actuation_force(q, u)
```
Comment (mstoelzle) [inline src/soromox/systems/pendulum.py:734]: Do you think it would make sense to add a `potential_force` method to `DynamicalSystem` that sums the elastic and the gravitational force?

File: src/soromox/systems/dynamical_system.py [hunk 4]
```diff
@@ -99,3 +102,36 @@ def resolve_upon_time(
         qs, qds = jnp.split(y_out, 2, axis=1)
 
         return ts, qs, qds
+
+    def solve_statics(
```
Comment (mstoelzle) [inline src/soromox/systems/dynamical_system.py:106]: jit missing

File: src/soromox/systems/dynamical_system.py [hunk 5]
```diff
@@ -99,3 +102,36 @@ def resolve_upon_time(
...
+        Returns:
+            sol (optx.Solution): Solution object of the root finder.
+        """
+        if u is None:
+            u = jnp.zeros((self.num_actuators,))
+        if tau_ext is None:
+            tau_ext = jnp.zeros((self.num_links,))
+        
+        sol = vmap(
+            lambda fn, s, y0, args: optx.root_find(fn, s, y0, args, max_steps=256, throw=False),
```
Comment (mstoelzle) [inline src/soromox/systems/dynamical_system.py:133]: The `max_steps` needs to be a parameter of the function.

**Review by mstoelzle — CHANGES_REQUESTED** [review]
The reasons why I am currently a bit sceptical of including a statics solver in the SoRoMoX main library are the following:

- As seen by the discussion we are having on Teams, you can formulate the problem in quite a few different ways. You can assume `u` to be given and identify `q` and vice-versa. Also, you can add the mapping into task space, and, for example, assume the task space position to be given. Finally, you can also define the task space in infinite different ways.
- There are quite a lot of different ways (e.g., gradient descent, nonlinear least squares, closed-form solutions in some cases, forward rollouts, etc.) how you can solve this problem from an optimization point of view and the same strategy might not work well for a wide class of systems and system parameters. 

Alternatively, we could just add an example for solving the statics.

**Comment by mstoelzle** [comment]
Closing this PR for now as solving the statics is a very complex problems settings with not a single solution, but instead many possible approaches, each with various advantages and disadvantages. Therefore, it is better in my opinion to include the statics in examples instead of the core package.
